packages/rest/src/coercion/coerce-parameter.ts

```diff
@@ -172,9 +179,9 @@ function parseJsonIfNeeded(
 ): string | object | undefined {
   if (typeof data !== 'string') return data;
 
-  if (spec.in !== 'query' || spec.style !== 'deepObject') {
+  if (spec.in !== 'query' || (spec.in === 'query' && !spec.content)) {
```

Comment:
IIUC, this is introducing another breaking change. If I am reading the code right:

- Before this change, when a parameter was described with `style: deepObject`, we attempted to parse it as JSON.
- After this change, when a parameter is described with `style: deepObject`, we never parse the value as JSON and require the value to be encoded using deep query string.

While the new behavior is technically more correct, I am little bit concerned about the impact on the existing applications. I understand this does not apply to apps using `@param.query.object` decorator, which is probably the majority of the apps out there. However, it is still possible to specify `style: deepObject` encoding for parameters, for example using generic `@param` decorator or even using a design-first approach where the entire OpenAPI spec is provided from outside via `app.api()` method, see [Defining the API using design-first approach](https://github.com/strongloop/loopback-next/blob/4406bd3b9de0fa023896f7c97cb2c98157ca9b71/docs/site/Defining-the-API-using-design-first-approach.shelved.md).

If don't have strong opinions on whether preserve old (more lenient) behavior or introduce the new (more strict) handling. If we decide to be more strict, then please make sure this change is spelled out in the release notes (the commit message).

Reply:
On the second thought, I think now is a good time to introduce this breaking change, since we are already changing other things. It should be easy (and backwards compatible) to reintroduce support for JSON-encoded deep-object parameters in the future, assuming there is anybody asking for that feature.

So the only remaining step is to update the commit message and explain the breaking change.

Reply:
> * After this change, when a parameter is described with `style: deepObject`, we never parse the value as JSON and require the value to be encoded using deep query string.
@bajtos share your concern on the above , but as of now there is no decorator to annotate a parameter with `style:deepObject` so there is no way a customer can have a case of trying to send a complex json for a parameter marked with spec `deepObject`.

Reply:
> * After this change, when a parameter is described with `style: deepObject`, we never parse the value as JSON and require the value to be encoded using deep query string.

for the change in this behavior I have added the following in the commit footer:
``` 
This fix has also modified the coercion behaviour for `in:query`, `deep-object` query
parameters by removing json parsing for them and the expectation is to pass in only
string values. But since the `exploded` `deep-object` style definition for query params
has been removed, this behaviour remains just an internal source code aspect as of now.
```